rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,108 @@
+
+On-prem customers could work around this "locked out" state by running `tctl` locally on the auth server to create a new admin user.
+
+## Details
+
+> Context: SSO users
+>
+> Teleport has this feature called SSO auth. It allows for the addition of an SSO connector to a cluster and maps SSO users/user groups to teleport roles. Currently, it works this way:
+>
+> 1. Authorized teleport users can add SSO connector (eg. GitHub SSO https://goteleport.com/docs/setup/admin/github-sso)
+> 2. GitHub users that are members of a particular organization and the team can log in as teleport users with a defined role.
+> 3. Users will show up in the users' list in UI but they will disappear when their cert expires (even when we remove the SSO connector).
+> 4. SSO users can have full access.
+
+The solution here could be to introduce a new rule:
+
+- There should be at least _admin_
+
+This will ensure that:
+
+- Users won't be able to delete _admin_ user if there is only one _admin_. They would have to add a second one and then they can delete it.
+- Users won't be able to unassign the user/role editing capabilities from _admin_ user if there are only is only one _admin_. They would have to add a second one and then they can unassign.
+- Locking will be impossible on the SSO side since there will be always _admin_ on our side.
+
+Now we need to define what _admin_ means in our system. W can take 2 approaches:
+
+1. User with an immutable `editor` role is _admin_.
+2. User with a role that grants user/role editing capabilities is _admin_.
+
+From the user's perspective, the first approach is simpler. The error message saying: "You can't delete the last user with `editor` role." will be easier to understand than "You can't delete the last user with: list of role management capabilities".
+
+The other benefit is that we wouldn't need to define validation for editing roles (since the `editor` role would be immutable).
+
+The drawbacks here are:
+
+- we force users to have the `editor` role
+- complicated migration in case of updating cluster. It is still possible to delete/edit default roles so not all existing clusters will have default roles in the initial form. Maybe the `editor` role is used as a role with a different meaning than the initial one? We can't override existing roles. So the solution would be to add 3 new roles (instead of editing/re-adding: editor, auditor, access). This will ensure we don't break anything. But the problem will be users would have to assign manually a new _admin_ role to some users.
+
+I think `2.` approach is better since migration cost outweighs the benefits of the pros.
+
+### UI and behavior changes
+
+We need to change the UI to reflect the new policy. We need to add validation to all actions that can cause lock (in `WebUI` and `tctl`). The error messages should inform the user why this action is not allowed.
```

Comment:
We need to change the UI how? It's not clear to me what's in scope.

Reply:
I meant that we will display dedicated error messages to the user. This doesn't require frontend changes

Reply:
Let's clarify the paragraph, but sounds good.

Reply:
e295a5985e1d68b2170cda9c622d1d156244c88b

Reply:
Thanks!